[SPARK-32754][SQL][TEST] Unify to `assertEqualJoinPlans` for join reorder suites

wzhfy opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Now three join reorder suites(`JoinReorderSuite`, `StarJoinReorderSuite`, `StarJoinCostBasedReorderSuite`) all contain an `assertEqualPlans` method and the logic is almost the same. We can extract the method to a single place for code simplicity.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
To reduce code redundancy.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Covered by existing tests.

wzhfy committed:
unify assertEqualPlans for join reorder related suites

wzhfy committed:
fix style

SparkQA commented:
**[Test build #128088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128088/testReport)** for PR 29594 at commit [`58c153f`](https://github.com/apache/spark/commit/58c153f81a418ad8159dae49179952342db859b1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32713/
Test PASSed.

SparkQA commented:
**[Test build #128088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128088/testReport)** for PR 29594 at commit [`58c153f`](https://github.com/apache/spark/commit/58c153f81a418ad8159dae49179952342db859b1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128088/
Test FAILed.

wzhfy commented:
cc @cloud-fan

SparkQA commented:
**[Test build #128089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128089/testReport)** for PR 29594 at commit [`647c854`](https://github.com/apache/spark/commit/647c8541faa595880835cd335e2d3b5138e81a60).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32714/
Test PASSed.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala:
```diff
@@ -198,4 +201,23 @@ trait PlanTestBase extends PredicateHelper with SQLHelper { self: Suite =>
     }
   }
+
+  def assertEqualPlansForJoinReorder(
```
Can we create a base test suite for join reorder?

wzhfy replied:
yes I also think that's better. Do you think it's necessary to create a separate folder called "joinReorder" under the optimizer package?

wzhfy replied:
@cloud-fan Updated. I add a new folder "joinReorder" under the "optimizer" test package, and put 3 suites and the new base test in it.

dongjoon-hyun replied:
Ur, @cloud-fan and @wzhfy , why a new test trait is better than this? The function looks useful in general.

wzhfy committed:
refactor

SparkQA commented:
**[Test build #128092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128092/testReport)** for PR 29594 at commit [`d35e4d8`](https://github.com/apache/spark/commit/d35e4d8a08c0d7dd91b7bec877ce829eddc10abd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32717/
Test PASSed.

SparkQA commented:
**[Test build #128089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128089/testReport)** for PR 29594 at commit [`647c854`](https://github.com/apache/spark/commit/647c8541faa595880835cd335e2d3b5138e81a60).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128089/
Test PASSed.

SparkQA commented:
**[Test build #128092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128092/testReport)** for PR 29594 at commit [`d35e4d8`](https://github.com/apache/spark/commit/d35e4d8a08c0d7dd91b7bec877ce829eddc10abd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait JoinReorderPlanTestBase extends PlanTest `
  * `class JoinReorderSuite extends JoinReorderPlanTestBase with StatsEstimationTestBase `
  * `class StarJoinCostBasedReorderSuite extends JoinReorderPlanTestBase with StatsEstimationTestBase `
  * `class StarJoinReorderSuite extends JoinReorderPlanTestBase with StatsEstimationTestBase `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128092/
Test PASSed.

dongjoon-hyun reviewed (commented):
~The refactoring looks okay. I have only one question on the decision, https://github.com/apache/spark/pull/29594#discussion_r480324425.~

Never mind. The name `assertEqualPlans` was misleading to me, but now I see that it's technically `assertEqualJoinPlan`.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/joinReorder/JoinReorderPlanTestBase.scala:
```diff
@@ -0,0 +1,51 @@
+    assert(equivalentOutput(analyzed, expected))
+    assert(equivalentOutput(analyzed, optimized))
+
+    compareJoinOrder(optimized, expected)
```
Oh, got it.

wzhfy committed:
rename and move two methods

wzhfy commented:
> ~The refactoring looks okay. I have only one question on the decision, [#29594 (comment)](https://github.com/apache/spark/pull/29594#discussion_r480324425).~
> 
> Never mind. The name `assertEqualPlans` was misleading to me, but now I see that it's technically `assertEqualJoinPlan`.

@dongjoon-hyun I think `assertEqualJoinPlan` is a better name, thanks!

wzhfy commented:
@cloud-fan @dongjoon-hyun I have renamed the method `assertEqualPlans` to `assertEqualJoinPlan`, and moved `compareJoinOrder` and `sameJoinPlan` from `` to `` since they are only used in join reorder test suites.

SparkQA commented:
**[Test build #128126 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128126/testReport)** for PR 29594 at commit [`638bbf9`](https://github.com/apache/spark/commit/638bbf991d426dccd53a27999d66628f70fc76cd).

wzhfy commented:
@cloud-fan @dongjoon-hyun 
I have renamed the method `assertEqualPlans` to `assertEqualJoinPlan`, and moved `compareJoinOrder` and `sameJoinPlan` from `PlanTest` to `JoinReorderPlanTestBase` since they are only used in join reorder test suites.
Could you please take another look? Thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32750/
Test PASSed.

SparkQA commented:
**[Test build #128126 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128126/testReport)** for PR 29594 at commit [`638bbf9`](https://github.com/apache/spark/commit/638bbf991d426dccd53a27999d66628f70fc76cd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128126/
Test PASSed.

dongjoon-hyun commented:
Thank you so much, @wzhfy and all. Merged to master for Apache Spark 3.1.0 on December 2020.
